.github/workflows/pypi-package.yml

```diff
@@ -0,0 +1,59 @@
+---
+name: Build and maybe publish to PyPI
+
+on:
+  push:
+    branches: [main]
+    tags: ["*"]
+  pull_request:
+    branches: [main]
+  release:
+    types:
+      - published
+  workflow_dispatch:
+
+
+permissions:
+  contents: read
+
+
+jobs:
+  deploy:
```

Comment:
For extra security, use `environment: pypi` and `environment: testpypi` on the separate release jobs. You can then add protection on the `pypi` env in the repo settings to require a manual approval and a cool-down timer while `testpypi` wouldn't have those. With the env separation you can even use the same secret name for those envs. I ended up defaulting to `PYPI_API_TOKEN`.

Reply:
I have done this now – can it still work with the releases trigger? where does the review into play?

Reply:
There's nothing special about the relation of triggers and environments assigned to jobs. The environment will be used whenever it's set, regardless of the trigger. The secret access, though, might still be dependent on the event in this case, I'm not sure.